[GT-141]-pre-select site, services and endpoints

Sae126V opened the pull request:
**Case 1:** If the user goes to the sites and tries to add recent downtimes. Then the user should be able to see the site, the services, and the endpoints pre-selected. 
**Case 2:**  If the user goes to the services and tries to add recent downtimes. Then the user should see the specific service-related endpoints that are pre-selected.

"Resolves #229"

rowan04 commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,6 +158,21 @@ function draw(\User $user = null) {
+        $se = \Factory::getServiceService()->getService($_GET['se']);
+        $site = \Factory::getSiteService()->getSite($_GET['site']);
+        if(\Factory::getRoleActionAuthorisationService()->authoriseAction(\Action::EDIT_OBJECT, $site, $user)->getGrantAction() == FALSE){
+            throw new \Exception("You don't have permission over $site");
```
```suggestion
            throw new \Exception("You don't have permissions over $site");
```

makes more grammatical sense?

rowan04 replied:
im not sure, permissions seems to roll off the tongue better

Sae126V replied:
Based on the code above, it is authorized only on the service ID and site level. So, If you think it is better, I am open and happy for the discussion, and we can create a tech debt for it to discuss and resolve.

Sae126V replied:
The reason I am saying is because we might need to change across the codebase.

rowan04 commented on htdocs/web_portal/controllers/downtime/view_endpoint_tree.php:
```diff
@@ -19,31 +19,34 @@
  /*====================================================== */
-function getServiceandEndpointList() {
+ 
+ // This is to authenticate request, lookup a User object by user's principle ID  and decides user is an Admin or NOT.
```
```suggestion
 // This is to authenticate a request, lookup a User object by the user's principle ID and decide if the user is an Admin or not.
```

Sae126V replied:
Updated..

rowan04 commented on htdocs/web_portal/controllers/downtime/view_endpoint_tree.php:
```diff
@@ -19,31 +19,34 @@
     $site = \Factory::getSiteService()->getSite($_REQUEST['site_id']);
     $services = $site->getServices();
     $params['services'] = $services;
+    if(isset($_REQUEST['se'])) $params['se'] = $_REQUEST['se'];
```
Does this line work?

rowan04 replied:
cool

Sae126V replied:
Done

rowan04 commented on htdocs/web_portal/controllers/downtime/view_endpoint_tree.php:
```diff
@@ -19,31 +19,34 @@
     $params['portalIsReadOnly'] = portalIsReadOnlyAndUserIsNotAdmin($user);
+}
+
+function getServiceandEndpointList() {
```
should the "and" be camel cased, to be getServiceAndEndpointList()? would need to be changed elsewhere when it is called

Sae126V replied:
Updated. I think it is carry forwarded.  Nice catch

rowan04 commented on htdocs/web_portal/views/downtime/view_nested_endpoints_list.php:
```diff
@@ -1,23 +1,44 @@
-        echo "<option value=\"s" . $service->getId() . "\" id=\"" . $service->getId() . "\" SELECTED>" . '('.xssafe($service->getServiceType()->getName()).') '.xssafe($service->getHostName()) . "</option>";
+
+        if ($specificServiceEndpoints) {
+            if ($service->getId() == $specificServiceEndpoints)
```
also in if function on line 28

Sae126V replied:
Yep, ONLY If it is a one line statement.

rowan04 replied:
cool, I've never done that before so didn't know if it worked. I'll resolve this and similar comments

rowan04 replied:
ok for some reason i can't resolve it

gregcorbett replied:
If it only works without brackets for a one line statement, please add the brackets in.

ineilson replied:
It will fall foul of PSR12 standard without the curly brackets - https://www.php-fig.org/psr/psr-12/#51-if-elseif-else

gregcorbett replied:
I'm somewhat surprised codeclimate didn't raise this.

gregcorbett replied:
ah https://docs.codeclimate.com/docs/phpcodesniffer#standard

>To use the PSR12 standard, you'll need to specify the beta channel in your .codeclimate.yml:

Sae126V replied:
Updated. Added {} for single line statements in the source code. I will make sure to go through DO and DON'Ts next time.

Sae126V commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,6 +158,21 @@ function draw(\User $user = null) {
     }
 
+    // URL Mapping
+    // If the user wants to fetch a specific service endpoints, show only that service endpoints.
```
Cool.. Updated

rowan04 replied:
was it?

Sae126V replied:
I am not sure on how it is missed. I have updated now. Cheers

rowan04 replied:
no problem :)

Sae126V commented on htdocs/web_portal/controllers/downtime/view_endpoint_tree.php:
```diff
@@ -19,31 +19,34 @@
     $params['portalIsReadOnly'] = portalIsReadOnlyAndUserIsNotAdmin($user);
+}
+
+function getServiceAndEndpointList() {
```
This was changed as part of this commit 5f4a55ca13058e08883043339813ff09eb97bfbe.

Sae126V commented on htdocs/web_portal/views/downtime/view_nested_endpoints_list.php:
```diff
@@ -1,23 +1,44 @@
+        if ($specificServiceEndpoints) {
+            if ($service->getId() == $specificServiceEndpoints)
+                $selected = 'SELECTED';
+            else
```
Done,

gregcorbett commented:
I've marked some of the issues codeclimate has found as "Confirmed". These look like they are mostly stylist things. Could these, and only these be addressed? The other things codeclimate has found are beyond the scope of this PR.

We try and ensure all new GOCDB code follows the [PSR-12 standard](https://www.php-fig.org/psr/psr-12/).

rowan04 commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,103 @@ function draw(\User $user = null) {
-        }
+    // URL Mapping for `site` ONLY
+    elseif (isset($_GET['site'])) {
+       displaySiteEndpoints($user, $nowUtcDateTime);
```
```suggestion
        displaySiteEndpoints($user, $nowUtcDateTime);
```
indent only 7 spaces, not 8

rowan04 commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,103 @@ function draw(\User $user = null) {
-                if(\Factory::getRoleActionAuthorisationService()->authoriseAction(\Action::EDIT_OBJECT, $se->getParentSite(), $user)->getGrantAction() ){
-                    $ses[] = $se;
-                }
+      displayAllServiceEndpoints($user, $nowUtcDateTime);
```
```suggestion
        displayAllServiceEndpoints($user, $nowUtcDateTime);
```
indent only 6 spaces, not 8

gregcorbett commented:
> If you add "Resolves #229" to the PR description, GitHub should automatically link this PR to the GitHub issue.

Can you add this?

When trying to test case 2, the Service box did not appear so I was unable to submit a downtime. Does that happen for you @Sae126V?

![image](https://user-images.githubusercontent.com/2488355/236257483-b02108f1-4f6f-44b3-939c-522f2fadac1b.png)

gregcorbett commented:
Codeclimate has also identified some more low hanging fruit for issues to fix - I've marked them as "CONFIRMED".

Sae126V commented on htdocs/web_portal/views/downtime/add_downtime.php:
```diff
@@ -497,7 +498,19 @@ function getSitesServices(){
+
+            if (getServiceIDFromQueryParams) {
+                // NOTE: If you use template literal for string interpolation, i.e., `${}`. It has to be in one line.
+                DVET_Path = 'index.php?Page_Type=Downtime_view_endpoint_tree' +
```
Hi @gregcorbett , Regarding the issue #https://github.com/GOCDB/gocdb/pull/439#issuecomment-1534999143.

It seems the jQuery Engine is having an issue with the template literal. It is wrongly interpreting the string interpolation if it is multi-line i.e., `${}`. I have changed the code to multi-line because code climate identified a problem with the space limit for a statement. It was working before when it was a single line statement for template literal. I have modified the code to support multi-line. Thus, resolving the space issue. I spend some time doing testing and figured. It should be working fine now.

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,103 @@ function draw(\User $user = null) {
-        $params = array('ses' => $ses, 'nowUtc' => $nowUtcDateTime->format('H:i T'), 'selectAll' => true);
-        show_view("downtime/add_downtime.php", $params);
-        die();
+    // URL Mapping for `site` and `se`
+    elseif (isset($_GET['site']) && isset($_GET['se'])) {
+        displaySiteAndSeEndpoints($user, $nowUtcDateTime);
```
```suggestion
    // URL Mapping for `site` and `se`.
    // If the user wants to add a downtime to a specific SE, pre-select that SE.
    elseif (isset($_GET['site']) && isset($_GET['se'])) {
        displaySiteAndSeEndpoints($user, $nowUtcDateTime);
```

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,103 @@ function draw(\User $user = null) {
-        $params = array('ses' => $ses, 'nowUtc' => $nowUtcDateTime->format('H:i T'), 'selectAll' => true);
-        show_view("downtime/add_downtime.php", $params);
-        die();
+    // URL Mapping for `site` and `se`
+    elseif (isset($_GET['site']) && isset($_GET['se'])) {
```
```suggestion
    // URL Mapping for `site` and `se`.
    // If the user wants to add a downtime to a specific site and SE,
    // show only that site's SEs.
    elseif (isset($_GET['site']) && isset($_GET['se'])) {
```

Sae126V replied:
I have updated the comments to be more readable

gregcorbett commented on htdocs/web_portal/views/downtime/add_downtime.php:
```diff
@@ -497,7 +498,19 @@ function getSitesServices(){
+            const getServiceIDFromQueryParams = new URL(this.location.href).searchParams.get('se');
+
+            if (getServiceIDFromQueryParams) {
+                // NOTE: If you use template literal for string interpolation, i.e., `${}`. It has to be in one line.
```
```suggestion
                // NOTE: If you use template literal for string interpolation,
                // i.e., `${}`, it has to be in one line.
```

Sae126V replied:
Done

gregcorbett commented on htdocs/web_portal/views/downtime/view_nested_endpoints_list.php:
```diff
@@ -1,26 +1,54 @@
+}
 ?>
-</select>
+</select>
```
Can yo add the new line character here?

gregcorbett commented on htdocs/web_portal/views/downtime/add_downtime.php:
```diff
@@ -630,22 +643,4 @@ function getURLParameter(name) {
-
-
-
+</script>
```
Can yo add the new line character here?

gregcorbett commented:
Looks good, I've added some comments / suggestions.

Sae126V commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,103 @@ function draw(\User $user = null) {
-           throw new \Exception("You do not have permission over $se.");
-        }
+    // URL Mapping for `site` ONLY
+    elseif (isset($_GET['site'])) {
```
I have updated the comments to be more readable

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,28 +158,50 @@ function draw(\User $user = null)
+     * If the user wants to add a downtime to a specific SE,
+     * show only that service endpoints.
+     * 
+     * TODO - Remove the dead code.
```
Dead code where?

gregcorbett replied:
which method?

Sae126V replied:
`displaySeEndpoints($user, $nowUtcDateTime);`.

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,28 +158,50 @@ function draw(\User $user = null)
+     * Generic URL mapping.
+     * 
+     * User should be able to see all the service endpoints associated with the site selected.
+     * TODO - Default selection of the value in the `Select Affected Site` section should be consistent.
```
Which part of the code does this refer to?

Sae126V replied:
![SiteNames](https://github.com/GOCDB/gocdb/assets/127398882/205ac8b8-46ff-4638-aaa7-60ec29a9cb09)

Default value in the Select Affected site Section: "This requires a team design discussion"

gregcorbett replied:
Something to discuss at the next dev meeting :)

Sae126V commented on htdocs/web_portal/controllers/downtime/view_endpoint_tree.php:
```diff
@@ -19,36 +19,52 @@
+
+function getServiceAndEndpointList()
+{
+    $params = array_replace([], initAuthRequest());
```
My intension was to declare a global variable. When Every time we call initAuthRequest() or whenever browser refresher(With or without Cache).  We are 100% certain that the value will be up-to date if we use array_replace. And, NOT carry forwarding the old values. But, we no longer need it hence I have removed it.

rowan04 commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+    $authorisationService = \Factory::getRoleActionAuthorisationService();
+    $editAction = \Action::EDIT_OBJECT;
+
+    // drop the ses where the user does not have edit permissions over
```
doesnt really make sense?

gregcorbett replied:
looks like this has been removed.

rowan04 commented on htdocs/web_portal/views/downtime/add_downtime.php:
```diff
@@ -1,6 +1,6 @@
 <!--
 @author James McCarthy
-
+ 
```
```suggestion

```

gregcorbett replied:
looks like this has been done

Sae126V commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+     * URL Mapping for `site` and `se`.
+     *
+     * If the user wants to add a downtime to a specific site and SE,
+     * show only that site SEs.
```
As per our functionality, it is to show all service endpoints under that service for a site. I don't think Singular is a good option to go with here. What do you say?. As it implies, it is only one Site and one service endpoint that we are displaying, But it is not; we display all SEs for a site. It is for more readability.

rowan04 replied:
> As per our functionality, it is to show all service endpoints under that service for a site. I don't think Singular is a good option to go with here. What do you say?. As it implies, it is only one Site and one service endpoint that we are displaying, But it is not; we display all SEs for a site. It is for more readability.

we can discuss it at the dev meeting tomorrow if u want, i think im not quite understanding the wording and how its supposed to work

gregcorbett replied:
The comment is better now 😄

Sae126V commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+     * URL Mapping for `site` ONLY.
+     *
+     * If the user wants to add a downtime to a specific site,
+     * show all of the site SEs.
```
Same as above. At any given instance, there will be only one site that we will be displaying in the Select Specific site section for the above condition because the user has selected that specific site for adding downtime. It should be singular here.

rowan04 replied:
> Same as above. At any given instance, there will be only one site that we will be displaying in the Select Specific site section for the above condition because the user has selected that specific site for adding downtime. It should be singular here.

when 's is used, its to show possession, not plurality. So its just showing the SEs belong to the site, not that it will show multiple sites. or am i getting confused here. can discuss tomorrow

gregcorbett replied:
The Service endpoint vs Service Endpoint endpoint terminology in GOCDB is confusing. I've left a suggestion for a comment here that I think is a little clearer: https://github.com/GOCDB/gocdb/pull/439/files#r1282082077

Sae126V commented:
I have updated my comments for better readability and addressed the issues . The rest of the comments need a team inputs. We can discuss it at the dev meeting.

gregcorbett commented:
There's an odd side effect of this change when clicking "Add Downtime" from the left hand menu. Going to `/portal/index.php?Page_Type=Add_Downtime` page shouldn't pre select a site, but it does.

![image](https://github.com/GOCDB/gocdb/assets/2488355/223c0b8c-47d2-48b0-bba3-4901ff330f3c)

Also, as an GOCDB Admin - I should see every site within GOCDB in the "Select Affected Site" list (as per GOCDB 5.10.4 behaviour), but when running this change, I only see sites that I have a Role over in that list.

Could you look into both those things @Sae126V ?

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+    }
+}
+
+// If the user wants to fetch a specific site and service endpoint, show only that service endpoint's.
```
I don't think this change is relevent any more.

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+    $site = \Factory::getSiteService()->getSite($_GET['site']);
+    $ses = $site->getServices();
+
+    if (\Factory::getRoleActionAuthorisationService()->authoriseAction(\Action::EDIT_OBJECT, $site, $user)->getGrantAction() == FALSE) {
```
it has been factored out 😄

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
     }
+
+    $ses = $site->getServices();
+    $params = array('ses' => $ses, 'nowUtc' => $nowUtcDateTime->format('H:i T'), 'selectAll' => true);
```
this has been done

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+    die();
+}
+
+// If the user wants to add a downtime to a specific site, show all the site SEs
```
I don't think this change is relevent any more as the comment has been changed.

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+       throw new \Exception("You do not have permission over $se.");
+    }
+
+    //$ses = array($se);
```
looks like this has been removed.

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+}
+
+/**
+ * Retrieves all service endpoints based on the user's role.
```
I don't think this change is relevent any more as the comment has been changed.

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+    die();
+}
+
+// If the user doesn't want to add a downtime to a specific SE or site show all SEs
```
I don't think this change is relevent any more as the comment has been changed.

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+/**
+ * Retrieves all service endpoints based on the user's role.
+ * If the user is an admin, It returns all the SE's.
+ * Otherwise, it retrieves the service endpoints based on the user permissions.
```
I don't think this change is relevent any more as the comment has been changed.

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+
+    //$ses = array($se);
+    $ses = $site->getServices();
+    $params = array('ses' => $ses, 'nowUtc' => $nowUtcDateTime->format('H:i T'), 'selectAll' => true);
```
this appears to have been done

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+    if ($user->isAdmin()) {
+        return \Factory::getServiceService()->getAllSesJoinParentSites();
+    } else {
+        //$allSites = \Factory::getUserService()->getSitesFromRoles($user);
```
this appears to have been removed.

gregcorbett commented on htdocs/web_portal/controllers/downtime/add_downtime.php:
```diff
@@ -158,62 +158,149 @@ function draw(\User $user = null) {
+
+    return $ses;
+}
 ?>
```
looks like a new line has been added.

Sae126V commented:
Raised another PR with CLEAN code.
